sympy/matrices/matrices.py

```diff
@@ -2357,22 +2358,28 @@ def __ne__(self, other):
         return not self == other
 
     def _matrix_pow_by_jordan_blocks(self, num):
-        from sympy.matrices import diag, MutableMatrix
+        from sympy.matrices import diag, MutableMatrix, eye
         from sympy import binomial
 
         def jordan_cell_power(jc, n):
             N = jc.shape[0]
             l = jc[0, 0]
-            if l == 0 and (n < N - 1) != False:
-                raise ValueError("Matrix det == 0; not invertible")
-            elif l == 0 and N > 1 and n % 1 != 0:
-                raise ValueError("Non-integer power cannot be evaluated")
-            for i in range(N):
-                for j in range(N-i):
-                    bn = binomial(n, i)
-                    if isinstance(bn, binomial):
-                        bn = bn._eval_expand_func()
-                    jc[j, i+j] = l**(n-i)*bn
+            if l.is_zero:
+                if N == 1 and n.is_nonnegative:
+                    jc [0, 0] = l**n
+                elif not (n.is_integer and n.is_nonnegative):
+                    raise NonInvertibleMatrixError("Non-invertible matrix can only be raised to a nonnegative integer")
+                else:
+                    for i in range (N):
+                        for j in range (N):
+                            jc[i,j] = KroneckerDelta(j-i, n) if i<=j else 0
+            else:
+                for i in range(N):
+                    for j in range(N-i):
+                        bn = binomial(n, i)
+                        if isinstance(bn, binomial):
+                            bn = bn._eval_expand_func()
+                        jc[j, i+j] = l**(n-i)*bn
```

Comment:
The above recomputes binomial(n, i) many times. Same goes for the expression `l**(n-i)*bn`. For larger matrices that could be significant.

We can use banded here to replace this double loop and the one above for the KroneckerDelta case. See #17213 as a similar example.

Reply:
To use banded the function would need to be changed so it's not modifying the matrix in place but I think that's a good change to make anyway.

Reply:
I was under the impression this community put more weight in clarity of meaning over efficiency but if not so the sure I can optimize it.